### DISCUSSION TITLE
FTP client

### DISCUSSION THREAD

**Discussion by JPSvdW** [zephyrproject-rtos/zephyr#31442] [Q&A] [answered]
Hi, I want to add an FTP client to my Zephyr project.  I am using an nRF52840 with a Quectell BG-96 modem and want to send data to an FTP server.  The only library I can find in Zephyr is the TFTP library which does not have much functionality yet.  Has anyone been able to maybe use a third-party FTP client library in Zephyr?

**Comment by felipedantassilva** [comment]
I´m in trouble with the same thing. The Nordic SDK Connect have a lib called ftp client (see this link: https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/libraries/networking/ftp_client.html). But none example avaliable. I´m working about this now. If we have some updatings, let´s going putting here, ok? Best regards

**Reply by jukkar** [reply]
If you decide to implement ftp client functionality, patches are welcome in upstream for it.

**Comment by matheusmbar** [comment]
I have interest in using a FTP client on Zephyr as well.
The NRF SDK implementation works on Zephyr 4.3.0, after copying the related source files, adding them on CMakeLists and kconfigs.

https://github.com/nrfconnect/sdk-nrf/tree/main/subsys/net/lib/ftp_client

It depends only on Zephyr libs. I basically had to enable these:

```
CONFIG_NETWORKING=y
CONFIG_NET_SOCKETS=y
CONFIG_NET_TCP=y
```

There is a catch though, these files are published under **LicenseRef-Nordic-5-Clause**
available at:

[SDK-NRF license](https://github.com/nrfconnect/sdk-nrf/blob/main/LICENSE)

It states that:

> 4. This software, with or without modification, must only be used with a
>    Nordic Semiconductor ASA integrated circuit.

Is it possible to use this implementation on Zephyr? Could Nordic provide this with a more open license for use in any integrated circuit?

**Comment by rlubos** [comment, marked as answer]
We've decided to upstream our FTP client library (with some tweaks) from NCS, the PR can be found here: 
https://github.com/zephyrproject-rtos/zephyr/pull/104316

It includes the library itself and a FTP shell module that serves as a basis for the sample app.

**Comment by matheusmbar** [comment]
Thanks a lot for contributing with that @rlubos.

I've just seen that https://github.com/zephyrproject-rtos/zephyr/pull/104316 has been merged.
Didn't test that yet, but it seems like this issue could be closed.
